File: backend/app/schemas/imports.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class SourceFileRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    file_name: str
    file_path: str
    file_size: int
    source_kind: str
    region: str | None
    company_name: str | None
    file_hash: str | None
    uploaded_at: datetime


class ImportBatchSummaryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    batch_name: str
    status: str
    created_at: datetime
    updated_at: datetime
    file_count: int


class ImportBatchDetailRead(ImportBatchSummaryRead):
    source_files: list[SourceFileRead]


class DeleteImportBatchesInput(BaseModel):
    batch_ids: list[str] = Field(min_length=1)


class DeleteImportBatchesRead(BaseModel):
    deleted_count: int
    deleted_ids: list[str]
    missing_ids: list[str]


class HeaderMappingPreviewRead(BaseModel):
    raw_header: str
    raw_header_signature: str
    canonical_field: str | None
    mapping_source: str
    confidence: float | None
    candidate_fields: list[str]
    matched_rules: list[str]
    llm_attempted: bool
    llm_status: str
    rule_overrode_llm: bool


class FilteredRowPreviewRead(BaseModel):
    row_number: int
    reason: str
    first_value: str


class NormalizedPreviewRecordRead(BaseModel):
    source_row_number: int
    values: dict[str, object | None]
    unmapped_values: dict[str, object | None]
    raw_values: dict[str, object | None]
    raw_payload: dict[str, object | None]


class SourceFilePreviewRead(BaseModel):
    source_file_id: str
    file_name: str
    source_kind: str
    region: str | None
    company_name: str | None
    raw_sheet_name: str
    raw_header_signature: str
    normalized_record_count: int
    filtered_row_count: int
    unmapped_headers: list[str]
    header_mappings: list[HeaderMappingPreviewRead]
    filtered_rows: list[FilteredRowPreviewRead]
    preview_records: list[NormalizedPreviewRecordRead]


class ImportBatchPreviewRead(BaseModel):
    batch_id: str
    batch_name: str
    status: str
    source_files: list[SourceFilePreviewRead]


class ValidationIssueRead(BaseModel):
    normalized_record_id: str | None
    source_row_number: int
    issue_type: str
    severity: str
    field_name: str | None
    message: str


class SourceFileValidationRead(BaseModel):
    source_file_id: str
    file_name: str
    raw_sheet_name: str
    issue_count: int
    issues: list[ValidationIssueRead]


class BatchValidationRead(BaseModel):
    batch_id: str
    batch_name: str
    status: str
    total_issue_count: int
    source_files: list[SourceFileValidationRead]


class MatchRecordRead(BaseModel):
    normalized_record_id: str | None
    source_row_number: int
    person_name: str | None
    id_number: str | None
    employee_id: str | None
    employee_master_id: str | None
    match_status: str
    match_basis: str | None
    confidence: float | None
    candidate_employee_ids: list[str]


class SourceFileMatchRead(BaseModel):
    source_file_id: str
    file_name: str
    raw_sheet_name: str
    result_count: int
    results: list[MatchRecordRead]


class BatchMatchRead(BaseModel):
    batch_id: str
    batch_name: str
    status: str
    employee_master_available: bool
    employee_master_count: int
    blocked_reason: str | None
    total_records: int
    matched_count: int
    unmatched_count: int
    duplicate_count: int
    low_confidence_count: int
    source_files: list[SourceFileMatchRead]


class ExportArtifactRead(BaseModel):
    template_type: str
    status: str
    file_path: str | None
    error_message: str | None
    row_count: int


class BatchExportRead(BaseModel):
    batch_id: str
    batch_name: str
    status: str
    export_job_id: str | None
    export_status: str | None
    blocked_reason: str | None
    artifacts: list[ExportArtifactRead]
    completed_at: datetime | None

File: frontend/src/services/imports.ts
import type { ApiSuccessResponse } from './api';
import { LONG_RUNNING_REQUEST_TIMEOUT_MS, apiClient } from './api';

export interface ImportSourceFile {
  id: string;
  file_name: string;
  file_path: string;
  file_size: number;
  region: string | null;
  company_name: string | null;
  file_hash: string | null;
  uploaded_at: string;
}

export interface ImportBatchSummary {
  id: string;
  batch_name: string;
  status: string;
  created_at: string;
  updated_at: string;
  file_count: number;
  created_by_name?: string | null;
  normalized_record_count?: number | null;
}

export interface ImportBatchDetail extends ImportBatchSummary {
  source_files: ImportSourceFile[];
}

export interface HeaderMappingPreview {
  raw_header: string;
  raw_header_signature: string;
  canonical_field: string | null;
  mapping_source: string;
  confidence: number | null;
  candidate_fields: string[];
  matched_rules: string[];
  llm_attempted: boolean;
  llm_status: string;
  rule_overrode_llm: boolean;
}

export interface FilteredRowPreview {
  row_number: number;
  reason: string;
  first_value: string;
}

export interface NormalizedPreviewRecord {
  source_row_number: number;
  values: Record<string, unknown>;
  unmapped_values: Record<string, unknown>;
  raw_values: Record<string, unknown>;
  raw_payload: Record<string, unknown>;
}

export interface SourceFilePreview {
  source_file_id: string;
  file_name: string;
  region: string | null;
  company_name: string | null;
  raw_sheet_name: string;
  raw_header_signature: string;
  normalized_record_count: number;
  filtered_row_count: number;
  unmapped_headers: string[];
  header_mappings: HeaderMappingPreview[];
  filtered_rows: FilteredRowPreview[];
  preview_records: NormalizedPreviewRecord[];
}

export interface ImportBatchPreview {
  batch_id: string;
  batch_name: string;
  status: string;
  source_files: SourceFilePreview[];
}

export interface BulkDeleteImportBatchesResult {
  deleted_count: number;
  deleted_ids: string[];
  missing_ids: string[];
}

export interface CreateImportBatchInput {
  files: File[];
  batchName?: string;
  region?: string;
  companyName?: string;
}

export interface FetchImportBatchPreviewOptions {
  sourceFileId?: string;
}

export interface BatchDeletionImpact {
  batch_id: string;
  batch_name: string;
  record_count: number;
  match_count: number;
  issue_count: number;
}

export async function fetchBatchDeletionImpact(batchId: string): Promise<BatchDeletionImpact> {
  const response = await apiClient.get<ApiSuccessResponse<BatchDeletionImpact>>(
    `/imports/${batchId}/deletion-impact`,
  );
  return response.data.data;
}

export async function fetchImportBatches(): Promise<ImportBatchSummary[]> {
  const response = await apiClient.get<ApiSuccessResponse<ImportBatchSummary[]>>('/imports');
  return response.data.data;
}

export async function fetchImportBatch(batchId: string): Promise<ImportBatchDetail> {
  const response = await apiClient.get<ApiSuccessResponse<ImportBatchDetail>>(`/imports/${batchId}`);
  return response.data.data;
}

export async function deleteImportBatch(batchId: string): Promise<void> {
  await apiClient.delete(`/imports/${batchId}`);
}

export async function bulkDeleteImportBatches(batchIds: string[]): Promise<BulkDeleteImportBatchesResult> {
  const response = await apiClient.post<ApiSuccessResponse<BulkDeleteImportBatchesResult>>('/imports/bulk-delete', {
    batch_ids: batchIds,
  });
  return response.data.data;
}

export async function createImportBatch(input: CreateImportBatchInput): Promise<ImportBatchDetail> {
  const formData = new FormData();
  for (const file of input.files) {
    formData.append('files', file);
  }
  if (input.batchName?.trim()) {
    formData.append('batch_name', input.batchName.trim());
  }
  if (input.region?.trim()) {
    formData.append('regions', input.region.trim());
  }
  if (input.companyName?.trim()) {
    formData.append('company_names', input.companyName.trim());
  }

  const response = await apiClient.post<ApiSuccessResponse<ImportBatchDetail>>('/imports', formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
    timeout: LONG_RUNNING_REQUEST_TIMEOUT_MS,
  });
  return response.data.data;
}

export async function parseImportBatch(batchId: string): Promise<ImportBatchPreview> {
  const response = await apiClient.post<ApiSuccessResponse<ImportBatchPreview>>(`/imports/${batchId}/parse`, undefined, {
    timeout: LONG_RUNNING_REQUEST_TIMEOUT_MS,
  });
  return response.data.data;
}

export async function fetchImportBatchPreview(batchId: string, options: FetchImportBatchPreviewOptions = {}): Promise<ImportBatchPreview> {
  const response = await apiClient.get<ApiSuccessResponse<ImportBatchPreview>>(`/imports/${batchId}/preview`, {
    params: options.sourceFileId ? { source_file_id: options.sourceFileId } : undefined,
    timeout: LONG_RUNNING_REQUEST_TIMEOUT_MS,
  });
  return response.data.data;
}
